server/src-lib/Hasura/GraphQL/Execute/LiveQuery/Plan.hs

```diff
@@ -211,15 +227,61 @@ validateVariables pgExecCtx variableValues = do
       res <- liftIO $ runExceptT (runLazyTx' pgExecCtx tx)
       liftEither res
 
-    -- Explicitly look for the class of errors raised when the format of a value provided
-    -- for a type is incorrect.
+    -- Explicitly look for the class of errors raised when the format of a value
+    -- provided for a type is incorrect.
     dataExnErrHandler = mkTxErrorHandler (has _PGDataException)
 
--- -------------------------------------------------------------------------------------------------
+-- | Internal: Used to collect information about various parameters
+-- of a subscription field's AST as we resolve them to SQL expressions.
+data QueryParametersInfo
+  = QueryParametersInfo
+  { _qpiReusableVariableValues     :: !GV.ReusableVariableValues
+  , _qpiSyntheticVariableValues    :: !(Seq (WithScalarType PGScalarValue))
+  , _qpiReferencedSessionVariables :: !(Set.HashSet SessVar)
+  -- ^ The session variables that are referenced in the query root fld's AST.
+  -- This information is used to determine a cohort's required session
+  -- variables
+  } deriving (Generic)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid QueryParametersInfo)
+
+makeLenses ''QueryParametersInfo
+
+-- | Resolves an 'GR.UnresolvedVal' by converting 'GR.UVPG' values to SQL
+-- expressions that refer to the @result_vars@ input object, collecting information
+-- about various parameters of the query along the way.
+resolveMultiplexedValue
+  :: (MonadState QueryParametersInfo m)
+  => GR.UnresolvedVal -> m S.SQLExp
+resolveMultiplexedValue = \case
+  GR.UVPG annPGVal -> do
+    let GR.AnnPGVal varM _ colVal = annPGVal
+    varJsonPath <- case varM of
+      Just varName -> do
+        modifying qpiReusableVariableValues $ Map.insert varName colVal
+        pure ["query", G.unName $ G.unVariable varName]
+      Nothing -> do
+        syntheticVarIndex <- use (qpiSyntheticVariableValues . to length)
+        modifying qpiSyntheticVariableValues (|> colVal)
+        pure ["synthetic", T.pack $ show syntheticVarIndex]
+    pure $ fromResVars (PGTypeScalar $ pstType colVal) varJsonPath
+  GR.UVSessVar ty sessVar -> do
+    modifying qpiReferencedSessionVariables (Set.insert sessVar)
+    pure $ fromResVars ty ["session", T.toLower sessVar]
```

Comment:
I wonder if this really needs to be done statefully. It makes me a little sad to do it this way, because there’s really no reason this needs to be imperative. The assignment of synthetic variables *does* need to be done statefully, since we’re assigning unique indices, but that is not true for this.

What if we did this separately in `buildLiveQuery` plan in a more direct fashion? I think something like this would work:

```haskell
let referencedSessionVars = astUnresolved ^.. traverseQueryRootFldAST._UVSessVar._2
```

where `(^..)` is our generalized variant from `Control.Lens.Extended`.

Reply:
> It makes me a little sad to do it this way, because there’s really no reason this needs to be imperative

Yup! Collecting variables too, need not be done statefully. I took this approach so that I don't have to worry about whether those three traversals will result in a single traversal in the generated code.

Reply:
That seems highly unlikely to have a meaningful impact on performance given everything else going on in the pipeline. If you can measure *any* difference between the two approaches, I will eat my words! 🙂 But otherwise, I’d rather do the thing that’s easier to understand.